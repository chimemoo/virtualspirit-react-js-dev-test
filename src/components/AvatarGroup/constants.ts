import avatar1 from '../../assets/avatar1.webp'
import avatar2 from '../../assets/avatar2.webp'
import avatar3 from '../../assets/avatar3.webp'


const userList = [
  {
    name: 'Budi',
    avatar: avatar1,
  },
  {
    name: 'Budi Mulya',
    // avatar: avatar2,
  },
  {
    name: 'Meli Sulis',
    avatar: avatar3,
  },
  {
    name: 'Meri',
    avatar: avatar1,
  },
  {
    name: 'Adi Mulya',
    avatar: avatar2,
  },
  {
    name: 'Sule Sulis',
    avatar: avatar3,
  }
]

export {
  userList,
}