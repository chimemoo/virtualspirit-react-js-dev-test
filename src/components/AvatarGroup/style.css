.avatar-group {
  display: flex;
  overflow: hidden;
}

.avatar-group .avatar {
  overflow: clip;
}

.avatar-group .avatar:not(:first-child) {
  margin-left: -8px;
}