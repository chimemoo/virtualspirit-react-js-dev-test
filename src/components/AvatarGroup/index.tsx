import Avatar from "../Avatar"
import { AvatarSizeType, UserListType } from "../types"
import './style.css'

interface AvatarGroupProps {
  userList: UserListType[]
  maxLength: number
  size: AvatarSizeType
}

function AvatarGroup({
  userList,
  maxLength,
  size,
}: AvatarGroupProps) {
  const showedAvatar = maxLength < userList.length
    ? userList.slice(0, maxLength)
    : userList

  const showMorePlaceholder = maxLength < userList.length

  return (
    <div className="avatar-group" data-testid="avatar-group">
      {showedAvatar.map((item, i) => (
        <Avatar
          size={size}
          url={item.avatar}
          name={item.name}
          key={item.key}
        />
      ))}
      {showMorePlaceholder && (
        <div className={`avatar more ${size}`}>
          <span>{`+${userList.length - maxLength}`}</span>
        </div>
      )}
    </div>
  )
}

export default AvatarGroup
