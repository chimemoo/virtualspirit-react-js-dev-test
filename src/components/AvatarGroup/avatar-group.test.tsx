import { render, screen } from '@testing-library/react'
import renderer from 'react-test-renderer'
import { expect, test, describe } from 'vitest'
import AvatarGroup from '.'
import userList from './__mocks__/userlist'


function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON()
  expect(result).toBeDefined()
  expect(result).not.toBeInstanceOf(Array)
  return result as renderer.ReactTestRendererJSON
}

describe('AvatarGroup Component Test', () => {
  test('Avatar Group with max length 3 and size xs showed correctly', () => {
    const component = renderer.create(
      <AvatarGroup userList={userList} maxLength={3} size="xs" />
    )

    let tree = toJson(component)
    expect(tree).toMatchSnapshot()

    render(<AvatarGroup userList={userList} maxLength={3} size="xs" />)

    const elementCount = screen.getByTestId('avatar-group').childElementCount;
    expect(elementCount).toBe(4)
  })

  test('Avatar Group with max length 4 and size sm showed correctly', () => {
    const ComponentToRender = <AvatarGroup userList={userList} maxLength={4} size="sm" />
    const component = renderer.create(ComponentToRender)

    let tree = toJson(component)
    expect(tree).toMatchSnapshot()

    render(ComponentToRender)

    const elementCount = screen.getByTestId('avatar-group').childElementCount;
    expect(elementCount).toBe(5)
  })


  test('Avatar Group with maxLength same with number of user list showed correctly', () => {
    const newUserList = userList.slice(0, 5) // get only 5 user list
    const ComponentToRender = <AvatarGroup userList={newUserList} maxLength={5} size="sm" />
    const component = renderer.create(ComponentToRender)

    let tree = toJson(component)
    expect(tree).toMatchSnapshot()

    render(ComponentToRender)

    const elementCount = screen.getByTestId('avatar-group').childElementCount;
    expect(elementCount).toBe(5)
  })

  test('Avatar Group with maxLength above the number of user list showed correctly', () => {
    const newUserList = userList.slice(0, 3) // get only 3 user list
    const ComponentToRender = <AvatarGroup userList={newUserList} maxLength={5} size="sm" />
    const component = renderer.create(ComponentToRender)

    let tree = toJson(component)
    expect(tree).toMatchSnapshot()

    render(ComponentToRender)

    const elementCount = screen.getByTestId('avatar-group').childElementCount;
    expect(elementCount).toBe(3)
  })
})