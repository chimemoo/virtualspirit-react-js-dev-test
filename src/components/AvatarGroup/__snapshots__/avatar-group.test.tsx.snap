// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AvatarGroup Component Test > Avatar Group with max length 3 and size xs showed correctly 1`] = `
<div
  className="avatar-group"
  data-testid="avatar-group"
>
  <div
    className="avatar xs"
    data-testid="avatar-image"
  >
    <img
      alt="Budi"
      src="/@fs/assets/avatar1.webp"
    />
  </div>
  <div
    className="avatar placeholder xs"
    data-testid="avatar-placeholder"
  >
    <span>
      BM
    </span>
  </div>
  <div
    className="avatar xs"
    data-testid="avatar-image"
  >
    <img
      alt="Meli Sulis"
      src="/@fs/assets/avatar3.webp"
    />
  </div>
  <div
    className="avatar more xs"
  >
    <span>
      +5
    </span>
  </div>
</div>
`;

exports[`AvatarGroup Component Test > Avatar Group with max length 4 and size sm showed correctly 1`] = `
<div
  className="avatar-group"
  data-testid="avatar-group"
>
  <div
    className="avatar sm"
    data-testid="avatar-image"
  >
    <img
      alt="Budi"
      src="/@fs/assets/avatar1.webp"
    />
  </div>
  <div
    className="avatar placeholder sm"
    data-testid="avatar-placeholder"
  >
    <span>
      BM
    </span>
  </div>
  <div
    className="avatar sm"
    data-testid="avatar-image"
  >
    <img
      alt="Meli Sulis"
      src="/@fs/assets/avatar3.webp"
    />
  </div>
  <div
    className="avatar sm"
    data-testid="avatar-image"
  >
    <img
      alt="Meri"
      src="/@fs/assets/avatar1.webp"
    />
  </div>
  <div
    className="avatar more sm"
  >
    <span>
      +4
    </span>
  </div>
</div>
`;

exports[`AvatarGroup Component Test > Avatar Group with maxLength above the number of user list showed correctly 1`] = `
<div
  className="avatar-group"
  data-testid="avatar-group"
>
  <div
    className="avatar sm"
    data-testid="avatar-image"
  >
    <img
      alt="Budi"
      src="/@fs/assets/avatar1.webp"
    />
  </div>
  <div
    className="avatar placeholder sm"
    data-testid="avatar-placeholder"
  >
    <span>
      BM
    </span>
  </div>
  <div
    className="avatar sm"
    data-testid="avatar-image"
  >
    <img
      alt="Meli Sulis"
      src="/@fs/assets/avatar3.webp"
    />
  </div>
</div>
`;

exports[`AvatarGroup Component Test > Avatar Group with maxLength same with number of user list showed correctly 1`] = `
<div
  className="avatar-group"
  data-testid="avatar-group"
>
  <div
    className="avatar sm"
    data-testid="avatar-image"
  >
    <img
      alt="Budi"
      src="/@fs/assets/avatar1.webp"
    />
  </div>
  <div
    className="avatar placeholder sm"
    data-testid="avatar-placeholder"
  >
    <span>
      BM
    </span>
  </div>
  <div
    className="avatar sm"
    data-testid="avatar-image"
  >
    <img
      alt="Meli Sulis"
      src="/@fs/assets/avatar3.webp"
    />
  </div>
  <div
    className="avatar sm"
    data-testid="avatar-image"
  >
    <img
      alt="Meri"
      src="/@fs/assets/avatar1.webp"
    />
  </div>
  <div
    className="avatar sm"
    data-testid="avatar-image"
  >
    <img
      alt="Adi Mulya"
      src="/@fs/assets/avatar2.webp"
    />
  </div>
</div>
`;
